pulsar-client/src/test/java/org/apache/pulsar/client/impl/ReaderImplTest.java

```diff
@@ -64,13 +66,20 @@ public void clean() {
     }
 
     @Test
-    void shouldSupportCancellingReadNextAsync() {
+    void shouldSupportCancellingReadNextAsync() throws IllegalAccessException {
         // given
-        CompletableFuture<Message<byte[]>> future = reader.readNextAsync();
+        reader.readNextAsync();
         Awaitility.await().untilAsserted(() -> {
             assertTrue(reader.getConsumer().hasNextPendingReceive());
         });
 
+        ConsumerBase consumer = (ConsumerBase)
+                FieldUtils.readDeclaredField(reader, "consumer", true);
+        ConcurrentLinkedQueue<CompletableFuture<Message<byte[]>>>
+                pendingReceives = (ConcurrentLinkedQueue<CompletableFuture<Message<byte[]>>>)
+                FieldUtils.readField(consumer, "pendingReceives", true);
+        CompletableFuture<Message<byte[]>> future = pendingReceives.peek();
+
```

Comment:
This will break the `reader.readNextAsync().cancel()` behavior because users will not do as this test does. I think the reason is the future instance in the pending receives is not the same one that was returned to the user.

I have tried this one, it should work for this case.

```java
public CompletableFuture<Message<T>> readNextAsync() {
        CompletableFuture<Message<T>> originalFuture = consumer.receiveAsync();
        CompletableFuture<Message<T>> result = originalFuture.thenApply(msg -> {
            consumer.acknowledgeCumulativeAsync(msg)
                    .exceptionally(ex -> {
                        log.error("[{}][{}] acknowledge message {} cumulative fail.", getTopic(),
                                getConsumer().getSubscription(), msg.getMessageId(), ex);
                        return null;
                    });
            return msg;
        });
        CompletableFutureCancellationHandler handler = new CompletableFutureCancellationHandler();
        handler.attachToFuture(result);
        handler.setCancelAction(() -> originalFuture.cancel(false));
        return result;
    }
```

Reply:
Yes, this seems to be the better approach if we want to keep supporting the `ReaderImpl.readNextAsync().cancel()` behavior.


I see that the return future of `MultiTopicsReaderImpl.readNextAsync()` is not cancellable now. Do we want to make the same change , `CompletableFutureCancellationHandler handler` in `MultiTopicsReaderImpl.readNextAsync() ` too?

Reply:
I assume the current behavior of `multiTopicsConsumer.receiveAsync().cancel()` is invalid. I also fixed this part and added a test to cover `multiTopicsConsumer.receiveAsync().cancel()` too.